### PR TITLE
ElastiCache Terraform Module

### PR DESCRIPTION
**Pull request by ollieanwyll**
Connects to ministryofjustice/cloud-platform#304

**WHAT**

- Terraform module that creates an AWS ElastiCache Cluster. 

- Terraform module that creates an ElastiCache subnet group that deploys within the live-0 VPC.

**WHY**

- So teams can automate the creation of ElastiCache Clusters.

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -1,14 +1,19 @@
...
-resource "aws_elasticache_cluster" "example" {
-  cluster_id           = "cluster-example"
-  engine               = "redis"
-  engine_version       = ""
-  node_type            = "cache.m3.medium"
-  num_cache_nodes      = 1
-  parameter_group_name = "default.redis3.2"
-  port                 = 6379
+resource "random_id" "id" {
+  byte_length = 4
```
Comment (alkar) [inline main.tf]: Can we bump this up please to maybe 8 bytes? Feels too short :)

File: variables.tf [hunk 2]
```diff
@@ -37,4 +49,9 @@ variable "number_of_nodes" {
 variable "port" {
     description = "The port number of your ElastiCache Cluster. Redis uses '6379' and Memached uses '11211'. The default will be '6379' with Redis being the assumed engine"
     default = 6379
-}
\ No newline at end of file
+}
+
+variable "ec_subnet_groups" {
+    description = "A list of VPC subnet IDs."
+    default = ["subnet-7293103a", "subnet-7bf10c21", "subnet-de00b3b8"]
```
Comment (alkar) [inline variables.tf]: These are the public subnets for the VPC, can you change them to the private ones please?

**Comment by alkar** [comment]
@ollieanwyll the proposed solutions for removing the hardcoded subnet are best left for a separate PR since they also require the pipeline to be updated as well.

File: main.tf (hunk 1 above)
Comment (ollieanwyll) [reply main.tf]: 4 bytes of hex is 8 characters long. Still want more?

Comment (alkar) [reply main.tf]: Yes please! The `id` can be up to 20 characters so it fits - more randomess, less collisions.

**Comment by alkar** [comment]
Leaving [this here as a reference](https://github.com/terraform-providers/terraform-provider-aws/issues/4105): authentication for redis, pending terraform provider changes upstream

File: example/outputs.tf [hunk 3]
```diff
@@ -0,0 +1,3 @@
+output "endpoint" {
```
Comment (alkar) [inline example/outputs.tf]: Is there a reason why we don't export the entire dict here?
For multi-node clusters we're losing information this way (nodes 2...n) and I think constructing an endpoint presumes that client libraries consume this in a certain way. Should we not leave this up to the user to consume as the see fit?

File: variables.tf [hunk 4]
```diff
@@ -1,40 +1,57 @@
...
-    default = 6379
-}
\ No newline at end of file
+  description = "The port number of your ElastiCache Cluster. Redis uses '6379' and Memached uses '11211'. The default will be '6379' with Redis being the assumed engine"
+  default     = 6379
+}
+
+variable "ec_subnet_groups" {
+  description = "A list of VPC subnet IDs."
+  default     = ["subnet-7293103a", "subnet-7bf10c21", "subnet-de00b3b8"]
```
Comment (alkar) [inline variables.tf]: As mentioned before, the subnets here are wrong - we should be using the private subnets of the cluster. Can you please adjust here and across examples & docs?

File: example/outputs.tf (hunk 3 above)
Comment (razvan-moj) [reply example/outputs.tf]: dict won't work as a k8s secret, it expects a string; also the client library gets the address of the other nodes once connected via API

**Pull request merged**